feat: add the ability to add custom regex to route params

vsergiu93 opened the pull request:
This PR adds the possibility to add custom regex constrains on dynamic route segments

### Syntax:
`/blog/{id:\d{1,9}}`
`/blog/{category}/{type:article|news}`

This PR only adds the capability to add custom regex, but there is no developer feedback, I mean at some point we should add some validation on this custom regexes so that we add some DX value, but I think this could be added as future improvements.

brendt commented on src/Tempest/Http/src/Route.php:
```diff
@@ -30,9 +30,12 @@ public function __construct(
-            '#\{(\w+)}#',
-            '([^/]++)',
+        $matchingRegex = preg_replace_callback(
+            '#\{(\w+)(?::([^}]+))?\}#',
```
Something I find helpful for clarity is to use named capture groups: `(?<name>?::([^}]+))`

Then you can access it via `$matches['name']`

brendt commented on src/Tempest/Http/src/GenericRouter.php:
```diff
@@ -165,13 +165,13 @@ private function resolveParams(Route $route, string $uri): ?array
         }
 
-        $tokensResult = preg_match_all('#\{\w+}#', $route->uri, $tokens);
+        $tokensResult = preg_match_all('#\{(\w+)(?::[^}]*)?}#', $route->uri, $tokens);
```
I'd use named capture groups for clarity (see next comment)

vsergiu93 commented:
@brendt I feel like I should add more test cases but not sure where, I'm not sure if I'm missing some edge cases or not.

aidan-casey commented:
@vsergiu93 - I have _not_ read through the code here yet, however, I'm curious... Have you benchmarked this at all?

vsergiu93 commented:
Hey @aidan-casey, no, I guess I should right? If I'm not mistaken there is some repo already prepared with lots of controllers

aidan-casey commented:
Yeah. There is one [here](https://github.com/tempestphp/tempest-benchmark). I _personally_ haven't checked in on it for a while, so while I think it is up-to-date, not sure. 🙂 

@brendt - I wonder if we could update CI to automatically test benchmarks.

vsergiu93 commented:
I think doing some benchmarks in the CI could be great.

vsergiu93 commented:
@aidan-casey I'll try to look in to running some benchmarks tomorrow morning

vsergiu93 commented:
I managed to run some benchmarks: 

## Setup
I cloned the [tempest-benchmark](https://github.com/tempestphp/tempest-benchmark) repo and did some changes to `setup/setup.php` file, I'm generating only 100 items per type (this is because I was getting that error above with the original number) and I also added an `return new Ok('test');` response to the templates (where applicable) 

For the test itself I used the example code from here https://github.com/tempestphp/tempest-framework/pull/175 and modified it to make it work.

```php 
<?php

declare(strict_types=1);

namespace App\Bench;

use PhpBench\Attributes\Revs;
use Tempest\Clock\Clock;
use Tempest\Clock\GenericClock;
use Tempest\Console\Testing\ConsoleTester;
use Tempest\Container\Container;
use Tempest\Container\GenericContainer;
use Tempest\Core\AppConfig;
use Tempest\Core\Kernel;
use Tempest\Core\Tempest;
use Tempest\Framework\Testing\Http\HttpRouterTester;
use Tempest\Http\GenericRequest;
use Tempest\Http\GenericRouter;
use Tempest\Http\Mappers\RequestToPsrRequestMapper;
use Tempest\Http\Method;
use Tempest\Http\Request;
use Tempest\Http\RouteConfig;
use Tempest\Http\RouteDiscovery;

use Tempest\Http\Router;

use function Tempest\map;

class GenericRouterBench
{
    protected string $root;

    protected bool $discoveryCache = false;

    /** @var \Tempest\Core\DiscoveryLocation[] */
    protected array $discoveryLocations = [];

    protected AppConfig $appConfig;

    protected Kernel $kernel;

    protected Container $container;

    protected ConsoleTester $console;

    protected HttpRouterTester $http;

    protected GenericRequest $genericRequest;

    public function __construct()
    {
        $this->root ??= __DIR__ . '/../../';

        $this->kernel ??= new Kernel(
            root: $this->root,
            discoveryLocations: $this->discoveryLocations,
            discoveryCache: $this->discoveryCache,
        );

        $this->container = $this->kernel->container;

        $this->console = $this->container->get(ConsoleTester::class);
        $this->http = $this->container->get(HttpRouterTester::class);

        // first static route
        $this->genericRequest = new GenericRequest(Method::GET, '/admin/controller_001', []);
        // last static route
        // $this->genericRequest = new GenericRequest(Method::GET, '/users/controller_100', []);
        // first dynamic route
        //$this->genericRequest = new GenericRequest(Method::GET, '/books/xyz', []);
        // last dynamic route
        //$this->genericRequest = new GenericRequest(Method::GET, '/test/a/b', []);

        $this->container->singleton(Request::class, fn () => $this->genericRequest);
        $this->container ->singleton(GenericRequest::class, fn () => $this->genericRequest);
    }

    #[Revs(10_000)]
    public function benchDispatch(): void
    {
        //$request = map($this->genericRequest)->with(RequestToPsrRequestMapper::class);
        /** @var Router $router */
        $router = $this->container->get(Router::class);

        $router->dispatch($this->genericRequest);
    }
}
```


### Before my changes:
1.  **First static route** 
![image](https://github.com/user-attachments/assets/226f1593-361b-440c-8d09-9285f9cf505d)

###  After my changes:
1.  **First static route** 
![image](https://github.com/user-attachments/assets/33f60c6c-9afa-480b-9212-1fc3757fcce1)

@aidan-casey Let me know what you think, I have to be honest, I'm a bit lost here, any ideas on specific test cases?

brendt commented:
I actually think `preg_match(): Compilation failed: regular expression is too large at offset 290350` is a pretty significant issue. My guess is you're running into it because there's now more regex added to the routes. 

I do want to look into it, but right now I have other things higher on my priority list :/

vsergiu93 commented:
I have to mention that I was running in to that when generating 1k items per type, so that is about 3k controllers which each have like 4 or 5 routes (I'm talking here of the setup code that exists in tempest-benchmark). For sure allowing custom regex per params could increase the big regex size but that is only if you have routes with custom regex otherwise the code will jut put what we had before for each segment and I haven't really modified the templates to use custom regex params.

I think this issue theoretically exists because of the  approach taken here on routing and generating one big regex, but, at the same time I think that there is a small chance someone has build or plans to build a project containing around 4-5k routes (the biggest that I worked on was 600+ routes built with laravel).

@brendt I think the issue regex being to large is not connected to my changes, or better put, is not an issue that was introduced with this changes, the issue is there either way, it may be that with my changes you need fewer routes to hit it but you will hit it either way in the end.
I think a separate issue should be created with this and research some possible solutions to chunk the big regex or something similar.

Btw, if you want we can put this back to draft or feel free to close it if you want.

aidan-casey reviewed (commented):
@vsergiu93 - First off, good work on this PR and thank you for digging into this! You've inspired me to keep looking at the router before `v1` as there are some more optimizations we still haven't put together here.

Without those other optimizations in place, I'm not sure we will be able to improve performance too much. Overall, this looks pretty good. Just left a few comments.

Out of curiosity, do you know where the decrease in performance from your benchmarks is coming from?

aidan-casey commented:
> @brendt I think the issue regex being to large is not connected to my changes, or better put, is not an issue that was introduced with this changes, the issue is there either way, it may be that with my changes you need fewer routes to hit it but you will hit it either way in the end.

Side note, you are completely correct here, @vsergiu93. We were supposed to revisit the router to implement chunking, grouping of routes, and nested matching. None of that has happened yet, so I will open an issue for it to look into.

aidan-casey commented on src/Tempest/Http/src/Route.php:
```diff
@@ -28,12 +29,12 @@ public function __construct(
+        // Routes can have parameters in the form of "/{PARAM}/" or /{PARAM:CUSTOM_REGEX},
+        // these parameters are replaced with a regex matching group or with the custom regex
+        $matchingRegex = (string)str($this->uri)->replaceRegex(
+            '#\{(?<name>\w+)(?::(?<regex>[^}]+))?\}#',
```
@vsergiu93 - I think there is a bug here. If you can fix that, I will merge this and we can address the other pieces in future PRs.

vsergiu93 replied:
@aidan-casey I'll have a look thanks

vsergiu93 replied:
> What happens with something someone shouldn't do like:
> 
> ```regexp
> /user/{name: {name}}
> ```
 
Well, you'll get some malformed $matchingRegex, `/user/( {name)}` this is obviously incorrect, and I think here we need to see how to approach this. From DX pov I would say it would be nice to throw some exceptions, but that means we need to validate the custom regex therefore performance might be impacted (we could maybe conditionally do this check only in development and disable this in prod).

I guess we could check how other frameworks handle this part, see what kind of checks they do if any, if there are any restrictions in place.

Also @aidan-casey, I think I would like to write some unit tests for src/Tempest/Http/src/Route.php, I mean with the integrations tests is quite cumbersome to test this part with custom regex properly. Where should I put this tests?

aidan-casey replied:
> and I think here we need to see how to approach this

My inclination would be to just ignore nested brackets. See [here](https://github.com/nikic/FastRoute/blob/master/src/RouteParser/Std.php) for inspiration.

> I think I would like to write some unit tests for src/Tempest/Http/src/Route.php

Brilliant idea! I would place these in `src/Tempest/Http/tests/RouteTest.php`.

vsergiu93 replied:
@aidan-casey Thanks for the link, yeah, as I was suspecting already, ideally some checks need to be put in place, like 
not allowing users to have something like this `/{lang:(en|de)}` or this `/foo/{test}/{test:\d+}`. 
I guess it would make sense to add some checks and throw some exceptions to improve DX, but not sure if we should do it in this PR.

I did some improvements (basically I reused some of the regex from the packaged you linked), you could check the unit test input and expected for different formats and let me know if is ok. 

One more thing I noticed that FastRoute allows for route segment name to be surrounded  by spaces `/foo/{ param : \d{1,9} }` , I did add a trim for the custom regex, do you think we should handle this case as well?

aidan-casey replied:
Let's get this merged and revisit in some follow-up PRs. I think these deserve some broader discussion.

vsergiu93 commented on src/Tempest/Http/src/Route.php:
```diff
@@ -28,12 +29,12 @@ public function __construct(
+        // these parameters are replaced with a regex matching group or with the custom regex
+        $matchingRegex = (string)str($this->uri)->replaceRegex(
+            '#\{(?<name>\w+)(?::(?<regex>[^}]+))?\}#',
+            fn ($matches) => isset($matches['regex']) ? '(' . $matches['regex'] . ')' : '([^/]++)'
```
Apparently yes, `$matches` will not always contain a the `regex` 
key, basically you'll get an error for 
`Undefined index: regex ...`. I did something though, let me know if you like it.

vsergiu93 commented on src/Tempest/Http/src/GenericRouter.php:
```diff
@@ -170,13 +172,13 @@ private function resolveParams(Route $route, string $uri): ?array
         }
 
-        $tokensResult = preg_match_all('#\{\w+}#', $route->uri, $tokens);
+        $tokensResult = preg_match_all('#\{(?<names>\w+)(?::[^}]*)?}#', $route->uri, $tokens);
```
I will do some reading tomorrow on this, I don't mind removing them :)

brendt replied:
Fair enough

vsergiu93 replied:
I think the performance impact is minimal, nevertheless I decided to remove the named matching. Also it annoyed me that every time I had to change the regex for capturing the param's custom regex I had to do it in 3 places, not sure if my approach is right though, let me know what you think.

aidan-casey replied:
Thanks!

aidan-casey commented:
@vsergiu93 - I think this is good enough to merge. Let's address the discussion points in a follow-up issue / PR.
